cmake/CMakeLists.txt

```diff
@@ -659,8 +659,16 @@ function(onnxruntime_set_compile_flags target_name)
         if(TARGET ${onnxruntime_external_lib})
           get_target_property(onnxruntime_external_lib_include_dirs ${onnxruntime_external_lib} INTERFACE_INCLUDE_DIRECTORIES)
           foreach(onnxruntime_external_lib_include_dir IN LISTS onnxruntime_external_lib_include_dirs)
-            # TODO: onnxruntime_external_lib_include_dir is a cmake generator expr. Very surprising it can fit in at here.
-            target_compile_options(${target_name} PRIVATE "$<$<NOT:$<COMPILE_LANGUAGE:CUDA>>:/external:I${onnxruntime_external_lib_include_dir}>")
+            if(onnxruntime_external_lib_include_dir MATCHES "^\\$")
+              if(onnxruntime_external_lib_include_dir MATCHES "^\\$<BUILD_INTERFACE:([^>]+)>$")
+                string(REGEX REPLACE "^\\$<BUILD_INTERFACE:([^>]+)>$" "\\1" onnxruntime_external_lib_include_dir_cmake "${onnxruntime_external_lib_include_dir}")
+                cmake_path(NATIVE_PATH onnxruntime_external_lib_include_dir_cmake NORMALIZE onnxruntime_external_lib_include_dir_native)
+                target_compile_options(${target_name} PRIVATE "$<$<NOT:$<COMPILE_LANGUAGE:CUDA>>:/external:I${onnxruntime_external_lib_include_dir_native}>")
```

Comment:
is the problem with nested generator expressions? would something like https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:GENEX_EVAL help?

Reply:
Some of the include dirs are empty. For example, it may have things like "$<BUILD_INTERFACE:>", which will be evaluated as an empty string. Then this "/external:I" probably doesn't have a value, and it may misuse another thing after it as the value.

Reply:
I see, thanks. might be useful to include a comment explaining why we need to do this